pkg/identity/basicallocator/basic_allocator.go

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package basicallocator
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/idpool"
+)
+
+// BasicIDAllocator is used for managing a range of identities (unsigned
+// integers) in a very simple way, by allocating specified or random identities,
+// and returning them to the available pool.
+// Concurrency control (mutex locks) is built into IDPool.
```

Comment:
Can you explain what this library provides on top of `idpool`? I see that `IDPool` already has minimum, maximum ID values and random allocation.

Reply:
I see conversion to interfaces based on `error` and some renaming (+ hiding of complexity of the underlying type). Seems like the underlying structure was intended to be as error-free as possible and just have clear idempotent behaviour (mostly with boolean results rather than errors).

I guess.. If we as a project think that the underlying library should just use errors, then maybe we should change `IDPool` for everyone including using typed errors? I see that `idpool` is used by around four packages currently, so if this is broadly beneficial to the tree then why not use this pattern everywhere? Otherwise I'm not sure whether it's worth creating this intermediate layer when the intent is only to wrap this in yet one more type for more opinionated numeric allocation.

Reply:
`BasicIDAllocator` is intended to be used both cilium-operator (global identity management) and cilium-agent (temporary identity management). Both only require the simplest form of allocation, which is already implemented in `IDPool`. `BasicIDAllocator` just creates reuses this implementation and provides a simple interface to it.

I don't know if moving `IDPool` to return errors is correct, nor if we should rework it regardless of this change.

Wrapping `IDPool` instead of using it directly here for identity management is a good choice, because it's more accurate and simple, than handling `IDPool` in two places exactly the same way:
- `AllocateRandom()` instead of `AllocateID()`. Maybe renaming `IDPool`'s method would be good.
- `Allocate()` instead of `Remove()`.
- `ReturnToAvailablePool()` instead of `Insert()`.

Returning errors is just more convenient way to handle failed operations, because the operations are expected to fail only if something is actually wrong.